File: web/src/TrustRegistryService.ts
import ServiceBase from "./ServiceBase";
import {
  AddFrameworkRequest,
  AddFrameworkResponse,
  FetchDataRequest,
  FetchDataResponse,
  GetMembershipStatusRequest,
  GetMembershipStatusResponse,
  RegisterMemberRequest,
  RegisterMemberResponse,
  RemoveFrameworkRequest,
  RemoveFrameworkResponse,
  SearchRegistryRequest,
  SearchRegistryResponse,
  ServiceOptions,
  TrustRegistryDefinition,
  UnregisterMemberRequest,
  UnregisterMemberResponse,
} from "./proto";

import type { Client as BrowserClient } from "nice-grpc-web";

export class TrustRegistryService extends ServiceBase {
  client: BrowserClient<typeof TrustRegistryDefinition>;

  constructor(options?: ServiceOptions) {
    super(options);

    this.client = this.createClient(TrustRegistryDefinition);
  }

  public async registerMember(
    request: RegisterMemberRequest
  ): Promise<RegisterMemberResponse> {
    return this.client.registerMember(request, {
      metadata: await this.getMetadata(
        RegisterMemberRequest.encode(request).finish()
      ),
    });
  }

  public async unregisterMember(
    request: UnregisterMemberRequest
  ): Promise<UnregisterMemberResponse> {
    return this.client.unregisterMember(request, {
      metadata: await this.getMetadata(
        UnregisterMemberRequest.encode(request).finish()
      ),
    });
  }

  public async getMembershipStatus(
    request: GetMembershipStatusRequest
  ): Promise<GetMembershipStatusResponse> {
    return this.client.getMembershipStatus(request, {
      metadata: await this.getMetadata(
        GetMembershipStatusRequest.encode(request).finish()
      ),
    });
  }

  public async searchRegistry(
    request: SearchRegistryRequest = SearchRegistryRequest.fromPartial({
      query: "SELECT * FROM c OFFSET 0 LIMIT 100",
    })
  ): Promise<SearchRegistryResponse> {
    return this.client.searchRegistry(request, {
      metadata: await this.getMetadata(
        SearchRegistryRequest.encode(request).finish()
      ),
    });
  }

  public async addFramework(
    request: AddFrameworkRequest
  ): Promise<AddFrameworkResponse> {
    return this.client.addFramework(request, {
      metadata: await this.getMetadata(
        AddFrameworkRequest.encode(request).finish()
      ),
    });
  }

  public async removeFramework(
    request: RemoveFrameworkRequest
  ): Promise<RemoveFrameworkResponse> {
    return this.client.removeFramework(request, {
      metadata: await this.getMetadata(
        RemoveFrameworkRequest.encode(request).finish()
      ),
    });
  }

  public async fetchData(
    request: FetchDataRequest
  ): Promise<AsyncIterable<FetchDataResponse>> {
    return this.client.fetchData(request, {
      metadata: await this.getMetadata(
        FetchDataRequest.encode(request).finish()
      ),
    });
  }
}

File: java/src/main/java/trinsic/services/TrustRegistryService.java
package trinsic.services;

import com.google.common.util.concurrent.ListenableFuture;
import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.DidException;
import trinsic.sdk.options.v1.Options;
import trinsic.services.trustregistry.v1.*;
import trinsic.services.trustregistry.v1.TrustRegistryGrpc;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Iterator;
import java.util.concurrent.ExecutionException;

public class TrustRegistryService extends ServiceBase {
  public TrustRegistryGrpc.TrustRegistryFutureStub stub;
  // We only use the blocking stub for server-side streaming call.
  public TrustRegistryGrpc.TrustRegistryBlockingStub stub2;

  public TrustRegistryService() {
    this(null);
  }

  public TrustRegistryService(Options.ServiceOptions.Builder options) {
    super(options);
    this.stub = TrustRegistryGrpc.newFutureStub(this.getChannel());
    this.stub2 = TrustRegistryGrpc.newBlockingStub(this.getChannel());
  }

  public ListenableFuture<AddFrameworkResponse> addFramework(
      AddFrameworkRequest request)
      throws InvalidProtocolBufferException, DidException {
    try {
      new URL(request.getGovernanceFrameworkUri());
    } catch (MalformedURLException e) {
      throw new IllegalArgumentException("invalid uri string", e);
    }
    return withMetadata(stub, request).addFramework(request);
  }

  public RemoveFrameworkResponse removeFramework(
      RemoveFrameworkRequest request)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    return withMetadata(stub, request).removeFramework(request).get();
  }

  public RegisterMemberResponse registerMember(
      RegisterMemberRequest request)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    return withMetadata(stub, request).registerMember(request).get();
  }

  public UnregisterMemberResponse unregisterIssuer(
      UnregisterMemberRequest request)
      throws InvalidProtocolBufferException, DidException, ExecutionException,
          InterruptedException {
    return withMetadata(stub, request).unregisterMember(request).get();
  }

  public ListenableFuture<GetMembershipStatusResponse> checkIssuerStatus(
      GetMembershipStatusRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub, request).getMembershipStatus(request);
  }

  public ListenableFuture<SearchRegistryResponse> searchRegistry()
      throws InvalidProtocolBufferException, DidException {
    return searchRegistry(SearchRegistryRequest.getDefaultInstance());
  }

  public ListenableFuture<SearchRegistryResponse> searchRegistry(
      SearchRegistryRequest request)
      throws InvalidProtocolBufferException, DidException {
    if (request.getQuery().isBlank())
      request =
          SearchRegistryRequest.newBuilder(request)
              .setQuery("SELECT * FROM c OFFSET 0 LIMIT 100")
              .build();
    return withMetadata(stub, request).searchRegistry(request);
  }

  public Iterator<FetchDataResponse> fetchData(
      FetchDataRequest request)
      throws InvalidProtocolBufferException, DidException {
    return withMetadata(stub2, request).fetchData(request);
  }
}
